src/logic/shift.rs

```diff
@@ -0,0 +1,886 @@
+                //this becomes faster at about this point
+                x.rotate_left(s);
+                return
+            }
+            //in order to avoid allocation, we have to "leapfrog" and wrap around the indexing.
+            let mut start = 0;
+            let mut lowest_nonstart = s;
+            loop {
+                let mut tmp0 = x[start];
+                let mut tmp1;
+                let mut i = start + s;
+                loop {
+                    tmp1 = x[i];
+                    x[i] = tmp0;
+                    i += s;
+                    if i >= x.len() {
+                        i -= x.len();
+                        if i == start {
+                            x[i] = tmp1;
+                            break
+                        }
+                        //we cannot just have a `lowest_nonstart = i` here without a conditional
+                        //(counterexample is x.len() = 16 and s = 6)
+                        if i < lowest_nonstart {
+                            lowest_nonstart = i;
+                        }
+                    }
+                    //Repeat with swapped temporaries in case the compiler cannot do it.
+                    //We do not need a second indexing variable to handle the source
+                    //digit, because that source variable is already stored in tmp0 or
+                    //tmp1.
+                    tmp0 = x[i];
+                    x[i] = tmp1;
+                    i += s;
+                    if i >= x.len() {
+                        i -= x.len();
+                        if i == start {
+                            x[i] = tmp0;
+                            break
+                        }
+                        if i < lowest_nonstart {
+                            lowest_nonstart = i;
+                        }
+                    }
+                }
+                //the leapfrog wraps around until it arrives back where it started, but
+                //not all digits have been shifted the first time if `(x.len() % digits) == 0`.
+                start += 1;
+                if start == lowest_nonstart {return}
+            }
+        }
+
+        //for `ApInt`s with a bit width that is not an integer multiple of `Digit`s, and a shift
+        //equal to or larger than a digit, `subdigit_rotate_left` and `digit_rotate_left` should
+        //be used followed by this correct the shift.
+        //assumes `(digits > 0) && (end_bits > 0)`
+        fn nonmultiple_rotate_correction(x: &mut [Digit], end_bits: usize, digits: usize, shift_bits: usize) {
+            //digits > 0, so the bits after the end_bits will always all be wrap around bits
+            let unshift = Digit::BITS - shift_bits;
+            let unbits = Digit::BITS - end_bits;
+            if shift_bits == 0 {
+                //this is difficult to explain, the best way is to look at a diagram
+                //[Digit::BITS is 8 here]
+                //01234567_89abcdef //all of x
```

Comment:
Please always use upper-hex letters if you want to represent something hex encoded. It is much more readable.